File: SBDemoApp/BeaconsViewController.m

//  BeaconsViewController.m
//  SensorbergSDK
//
//  Created by Andrei Stoleru on 12/01/16.
//  Copyright © 2016 Sensorberg GmbH. All rights reserved.
//

#import "BeaconsViewController.h"

#import <SensorbergSDK/SensorbergSDK.h>

#import <SensorbergSDK/NSString+SBUUID.h>

#import <tolo/Tolo.h>

#warning Enter your API key here
#define kAPIKey     @"000"

@interface BeaconsViewController () {
    NSMutableDictionary *beacons;
}

@end

static NSString *const kReuseIdentifier = @"beaconCell";

@implementation BeaconsViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    beacons = [NSMutableDictionary new];
    
    [[SBManager sharedManager] setApiKey:kAPIKey delegate:self];
    //
    [[SBManager sharedManager] requestLocationAuthorization:YES];
    [[SBManager sharedManager] requestNotificationsAuthorization];
    //
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    self.title = @"Beacons";
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return beacons.allValues.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:kReuseIdentifier forIndexPath:indexPath];
    
    cell.imageView.image = nil;
    
    if ([beacons.allValues[indexPath.row] isKindOfClass:[SBMBeacon class]]) {
        SBMBeacon *beacon = beacons.allValues[indexPath.row];
        NSString *proximityUUID = [[NSString hyphenateUUIDString:beacon.uuid] uppercaseString];
        if ([[SensorbergSDK defaultBeaconRegions] valueForKey:proximityUUID]) {
            cell.textLabel.text = [[SensorbergSDK defaultBeaconRegions] valueForKey:proximityUUID];
        } else {
            cell.textLabel.text = proximityUUID;
        }
        cell.detailTextLabel.text = [NSString stringWithFormat:@"Major:%i  Minor:%i", beacon.major, beacon.minor];
        
        
    } else if ([beacons.allValues[indexPath.row] isKindOfClass:[CBPeripheral class]]) {
        CBPeripheral *p = beacons.allValues[indexPath.row];
        
        cell.textLabel.text = p.name ? p.name : @"No Name";
        cell.detailTextLabel.text = p.identifier.UUIDString;
        
        cell.imageView.image = [self imageFromText:@"BLE"];
    }
    
    return cell;
}

#pragma mark - Table view delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}

#pragma mark - SensorbergSDK events

#pragma mark SBEventLocationAuthorization
SUBSCRIBE(SBEventLocationAuthorization) {
    if (event.locationAuthorization==SBLocationAuthorizationStatusAuthorized) {
            [[SBManager sharedManager] startMonitoring];
    } else {
        NSLog(@"Location Service OFF, monitoring doesn't work");
    }
}

#pragma mark SBEventBluetoothAuthorization
SUBSCRIBE(SBEventBluetoothAuthorization) {
    // You only need to ask for Bluetooth authorisation when working with GATT
}

SUBSCRIBE(SBEventNotificationsAuthorization) {
    if (!event.notificationsAuthorization) {
//        When notifications are not allowed
    }
}

#pragma mark SBEventRegionEnter
SUBSCRIBE(SBEventRegionEnter) {
    [beacons setValue:event.beacon forKey:event.beacon.tid];
    [self.tableView reloadData];
    //
}

#pragma mark SBEventRegionExit
SUBSCRIBE(SBEventRegionExit) {
    [beacons setValue:nil forKey:event.beacon.tid];
    [self.tableView reloadData];
}

#pragma mark SBEventRangedBeacon
SUBSCRIBE(SBEventRangedBeacon) {
    [beacons setValue:event.beacon forKey:event.beacon.tid];
}

#pragma mark SBEventPerformAction
SUBSCRIBE(SBEventPerformAction) {
    UILocalNotification *notification = [[UILocalNotification alloc] init];
    notification.alertTitle = event.campaign.subject;
    notification.alertBody = [NSString stringWithFormat:@"Name: %@\nBody: %@",event.campaign.subject,event.campaign.beacon.tid];
    notification.alertAction = [NSString stringWithFormat:@"%@",event.campaign.payload];
    //
    if (event.campaign.fireDate) {
        notification.fireDate = event.campaign.fireDate;
    } else {
        if ([[SBManager sharedManager] canReceiveNotifications]) {
            [[SBManager sharedManager] reportConversion:kSBConversionSuccessful forCampaignAction:event.campaign.uuid];
        } else {
            [[SBManager sharedManager] reportConversion:kSBConversionUnavailable forCampaignAction:event.campaign.uuid];
        }
    }
    //
    [[UIApplication sharedApplication] scheduleLocalNotification:notification];
    //
}

#pragma mark - Utility

-(UIImage *)imageFromText:(NSString *)text
{
    CGSize size  = [text sizeWithAttributes:nil];
    
    UIGraphicsBeginImageContextWithOptions(size,NO,0.0);
    
    [text drawAtPoint:CGPointZero withAttributes:nil];
    
    // transfer image
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

@end

File: SBDemoAppSwift/BeaconsViewController.swift
//
//  BeaconsViewController.swift
//  SensorbergSDK
//
//  Created by Andrei Stoleru on 14/07/16.
//  Copyright © 2016 Sensorberg GmbH. All rights reserved.
//

import UIKit

import SensorbergSDK

import SensorbergSDK.NSString_SBUUID

class BeaconsViewController: UITableViewController {
    
    var beacons = NSMutableArray()
    
    let cellIdentifier = "beaconCell"

    override func viewDidLoad() {
        super.viewDidLoad()

//TODO: Enter API key
        
        let kAPIKey = "<< !!! ENTER API KEY HERE !!! >>"

        SBManager.shared().setApiKey(kAPIKey, delegate: self)
        
        SBManager.shared().requestLocationAuthorization(true)
        
        beacons = [];
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        self.title = "Beacons"
    }

    // MARK: - Table view data source

    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return beacons.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier, for: indexPath)

        let beacon:SBMBeacon = beacons.object(at: (indexPath as NSIndexPath).row) as! SBMBeacon

        let proximityUUID:String = beacon.uuid.uppercased()
        
        let beaconID:String = SensorbergSDK.defaultBeaconRegions()![proximityUUID] as! String
        
        if (!beaconID.isEmpty) {
            cell.textLabel?.text = beaconID
        } else {
            cell.textLabel?.text = beacon.uuid
        }
        
        cell.detailTextLabel?.text = "Major: " + String(beacon.major) + " Minor: " + String(beacon.minor)

        return cell
    }
 
    /*
     
     The method names have to be in the form "on"+<Event name>
     So, `SBEventLocationAuthorization` becomes `onSBEventLocationAuthorization`
     The full list of events is available at
     
     */
    
    @objc public func onSBEventLocationAuthorization(_ event:SBEventLocationAuthorization) {
        //        print(event)
        SBManager.shared().startMonitoring()
    }
    
    @objc public func onSBEventPerformAction(_ event:SBEventPerformAction) {
        print(event)
    }
    
    @objc public func onSBEventRegionEnter(_ event:SBEventRegionEnter) {
        beacons.add(event.beacon)
        self.tableView.reloadData()
    }
    
    @objc public func onSBEventRegionExit(_ event:SBEventRegionExit) {
        beacons.remove(event.beacon)
        self.tableView.reloadData()
    }

}
